### PR TITLE
Query-ify Instance::resolve

### PR DESCRIPTION
**Pull request by Aaron1011**
Split off from #65989

Instance::resolve is now a wrapper for a new `resolve_instance` query.
This greatly improves performance on several benchmarks

### CODE REVIEW

**Comment by Aaron1011** [comment]
The `style-servo` benchmark seems to be completely missing from the perf run

**Comment by Mark-Simulacrum** [comment]
style-servo is currently broken (on master, too): "error: truncating cast: the value 4294967292 requires 32 bits but the target type is only 16 bits" (amongst others).

File: src/librustc/ty/instance.rs
Comment (nikomatsakis) [inline src/librustc/ty/instance.rs]: Hmm, this seems like a place where canonicalization would be appropriate. We don't expect inference variables, but we *do* use this (I believe) without having fully erased regions, right? In that case, canonicalizing might eliminate duplicate work.

**Review by nikomatsakis — COMMENTED** [review]
Thoughts?

File: src/librustc/ty/instance.rs
Comment (Aaron1011) [reply src/librustc/ty/instance.rs]: I agree. However, I think https://github.com/rust-lang/rust/pull/67717 is already doing canonicalization changes related to (among other things) `Instance::resolve`.

Comment (nikomatsakis) [reply src/librustc/ty/instance.rs]: Heh, I just came here from reviewing #67717 to point out that it is making changes to `Instance::resolve` that would require canonicalization.

Comment (Aaron1011) [reply src/librustc/ty/instance.rs]: @nikomatsakis: Would you like me to add canonicalization in this PR, or should it be done in a follow-up PR after #67717 is merged?

**Comment by Zoxc** [comment]
Is there something blocking this PR? It holds up https://github.com/rust-lang/rust/pull/67953.

**Comment by Aaron1011** [comment]
I assumed that this was waiting on `style-servo` to be fixed, so that we can do a full perf run.

**Comment by nikomatsakis** [comment]
I have been waiting until we land #67117 -- though if we want to land this first, perhaps some assertions that we don't have inference variables would suffice.

**Comment by Aaron1011** [comment]
@nikomatsakis: I assume you meant https://github.com/rust-lang/rust/pull/67717?

**Comment by nikomatsakis** [comment]
Yes. Sorry. :)

**Comment by Aaron1011** [comment]
Now that the servo benchmarks have been fixed, can I get another perf run.

**Comment by Aaron1011** [comment]
@Mark-Simulacrum: I've fixed the merge conflict

**Comment by Zoxc** [comment]
Is this blocked on something now?

**Comment by Aaron1011** [comment]
@Zoxc: Yes: https://github.com/rust-lang/rust/pull/68505

**Comment by Aaron1011** [comment]
@nikomatsakis: Now that https://github.com/rust-lang/rust/pull/68505 has been merged, this PR is unblocked and ready for review.

I think it would be good to get another perf run, since it's been over a month since the previous one.

**Comment by Aaron1011** [comment]
Actually, it looks like `Instance::resolve` will need to become a canonical query.

**Comment by Aaron1011** [comment]
It like erasing regions is sufficient, since `resolve_instance` already ends up erasing regions.

**Comment by Aaron1011** [comment]
It looks like we're getting the same wins as before - however, there appears to be a 1.2%-1.4% regression in `regression-31157-opt` under incremental compilation.

Given the large wins for incremental compilation on other benchmarks, I suspect that this will be a net improvement for most projects. 

@nikomatsakis: Do you think more work needs to be done to address the regression before this is merged?

**Comment by Centril** [comment]
> @nikomatsakis: Do you think more work needs to be done to address the regression before this is merged?

The question wasn't asked of me, but I think this PR is pretty clearly a win, and the regression-31157-opt is small enough that it could *almost* be within the noise. 👍 for landing this.

File: src/librustc/ty/instance.rs
Comment (nikomatsakis) [inline src/librustc/ty/instance.rs]: Hmm, I am concerned about the use of `erase_regions` here. It feels potentially surprising to me, particularly since the returned `Instance` type can embed types/regions from the input (which would now be erased). Do you know what the impact of this PR is without this call?

Note that a lot of the folks invoking this (e.g., during trans and other passes) will already have erased regions.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Can we at least measure the impact with `erase_regions` calls removed?

File: src/librustc/ty/instance.rs
Comment (Aaron1011) [reply src/librustc/ty/instance.rs]: If we don't erase regions, we'd need to make this a canonical query, as [this ends up getting invoked with region variables](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/22006/logs/68). I don't see a straightforward way of benchmarking the difference, since we'd be adding the cost of canonicalization.

Do any of the callers care about the regions in the returned `Instance`? It looks like `Instance::resolve` is called by CTFE, codegen, and some symbol-related code, none of which (I think) should care about the regions in the returned `Instance`.

Comment (Aaron1011) [reply src/librustc/ty/instance.rs]: Actually, it turns out that `Instance::resolve` already ends up erasing regions. We either end up calling `tcx.codegen_fulfill_obligation` (which erases regions):

https://github.com/rust-lang/rust/blob/dd67187965e136bff1ed05e035293441c60f0790/src/librustc_ty/instance.rs#L73

or we call `tcx.subst_and_normalize_erasing_regions`, which also erases:

https://github.com/rust-lang/rust/blob/dd67187965e136bff1ed05e035293441c60f0790/src/librustc_ty/instance.rs#L22

I think erasing regions before we invoke `tcx.resolve_instance` won't actually affect the output - we just end up erasing regions slightly earlier than we otherwise would.

Comment (eddyb) [reply src/librustc/ty/instance.rs]: MIR optimizations may use `Instance::resolve`, but yeah, no result of `Instance::resolve` should be flowing into borrow-checking.

**Comment by Aaron1011** [comment]
@Dylan-DPC: This should be `S-waiting-on-review` - I've responded to the most recent round of review comments.

**Comment by Dylan-DPC** [comment]
@Aaron1011 thanks. made the change (Btw you can change labels with rustbot if you don't have access :)

**Comment by Aaron1011** [comment]
@JohnCSimon: See my previous comment - this should be `S-waiting-on-review`

**Comment by Aaron1011** [comment]
@nikomatsakis: Are there any changes you'd like me to make, or should I just rebase this?

File: src/librustc_middle/query/mod.rs
Comment (eddyb) [inline src/librustc_middle/query/mod.rs:1261]: Just spotted this: whenever you have a `ty::ParamEnv` and some thing referring the to same parameters it adds bounds in scope of, like `(DefId, SubstsRef)` here, the right thing to do is to use `ty::ParamEnvAnd<(DefId, SubstsRef)>`.

Doing so should result in better caching (as the bounds in scope aren't kept if they're not needed by the `SubtsRef`), and is more semantically meaningful.

File: src/librustc_middle/ty/query/keys.rs
Comment (eddyb) [inline src/librustc_middle/ty/query/keys.rs:300]: I guess a strong hint is that you had to add a new `Key` impl: generally, you shouldn't need one.

File: src/librustc_middle/query/mod.rs
Comment (eddyb) [reply src/librustc_middle/query/mod.rs:1261]: Will do this in a drive-by commit.

**Pull request merged**
